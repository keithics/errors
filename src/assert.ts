import * as _ from 'lodash';

/**
 * Usage: assert(object,MyError,'optional message)
 * eg: assert(user,ArkValidationError,'User is not found)
 *  Must be used in conjunction with catchAsync
 *
 *
 export const profile = catchAsync(async (req: Request, res: Response) => {
    const user = await User.findById(req.user);
    assert(user, ErrorNotFound);
    return res.jsonp(user);
  });

 */

/**
 class T - placeholder
 **/
class T {}

/**
 * Checks if the first parameter is false, empty object, collection, map, or set. and returns error
 * @param value Object,string,number, boolean or string to be evaluated
 * @param error KError
 * @param message optionally a custom error message
 */
export const assert = (value, error: new (...any: unknown[]) => T, message = '') => {
  if (value === true || typeof value === 'number' || !_.isEmpty(value)) {
    return true;
  }
  throw new error(message);
};

/**
 * Generic Error for Ark
 */
export class KError extends Error {
  protected code: number;
}

/**
 * All other errors
 * @name ErrorOther
 */
export class ErrorOther extends KError {
  constructor(message) {
    super(message);
    this.name = 'ErrorOther';
    this.message = message || 'An Error Occurred';
    this.code = 500;
    Error.captureStackTrace(this, ErrorOther);
  }
}

/**
 * Errors for Validation
 */
export class ErrorValidation extends KError {
  constructor(message) {
    super(message);
    this.name = 'ErrorValidation';
    this.code = 422;
  }
}

/**
 * 404 not found error and when _id is not found in the database
 */
export class ErrorNotFound extends KError {
  constructor(message) {
    super(message);
    this.name = 'ErrorNotFound';
    this.code = 404;
    this.message = '404 Not Found';
    Error.captureStackTrace(this, ErrorNotFound);
  }
}

/**
 * Errors that will not send back as json header type
 * useful if the service is not an API
 * Can be use if the error responses are not content-type/json, eg: html responses
 */
export class ErrorTeapot extends KError {
  constructor(message) {
    super(message);
    this.name = 'ErrorTeapot';
    this.code = 418;
    this.message = message || 'ErrorTeapot';
    Error.captureStackTrace(this, ErrorTeapot);
  }
}

/**
 * Unauthorized Error
 */
export class ErrorInvalidToken extends KError {
  constructor(message) {
    super(message);
    this.name = 'ErrorInvalidToken';
    this.code = 401;
    this.message = 'JWT Token is invalid';
    Error.captureStackTrace(this, ErrorInvalidToken);
  }
}
