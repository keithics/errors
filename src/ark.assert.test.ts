import { arkAssert, ArkErrorNotFound, ArkErrorOther, ArkErrorValidation } from './ark.assert';

describe('Ark Error Handler Tests', () => {
  test('Validation Error Tests - should return error if object is null, false or an empty object', () => {
    expect(() => {
      return arkAssert(null, ArkErrorValidation, 'Object is required');
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      return arkAssert(false, ArkErrorValidation, 'Object is required');
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      return arkAssert({}, ArkErrorValidation, 'Object is required');
    }).toThrowErrorMatchingSnapshot();
  });

  test('Validation Error Tests - should not return an error if object is not empty or true', () => {
    expect(() => {
      return arkAssert(true, ArkErrorValidation, 'Object is required');
    }).toMatchSnapshot();

    expect(() => {
      return arkAssert({ name: 'name' }, ArkErrorValidation, 'Object is required');
    }).toMatchSnapshot();
  });

  test('404 Error Tests - should return error 404 if value is null.false or empty', () => {
    expect(() => {
      return arkAssert(null, ArkErrorNotFound);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      return arkAssert(false, ArkErrorNotFound);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      return arkAssert({}, ArkErrorNotFound);
    }).toThrowErrorMatchingSnapshot();
  });

  test('404 Error Tests - should NOT return error 404 if value is NOT null,false or empty', () => {
    expect(() => {
      return arkAssert(true, ArkErrorNotFound);
    }).toMatchSnapshot();

    expect(() => {
      return arkAssert({ name: 'name' }, ArkErrorNotFound);
    }).toMatchSnapshot();
  });

  test('500 Error Tests - should return error 500 if value is null.false or empty', () => {
    expect(() => {
      return arkAssert(null, ArkErrorOther);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      return arkAssert(false, ArkErrorOther);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      return arkAssert({}, ArkErrorOther);
    }).toThrowErrorMatchingSnapshot();
  });

  test('500 Error Tests - should NOT return error 500 if value is NOT null,false or empty', () => {
    expect(() => {
      return arkAssert(true, ArkErrorOther);
    }).toMatchSnapshot();

    expect(() => {
      return arkAssert({ name: 'name' }, ArkErrorOther);
    }).toMatchSnapshot();
  });
});
