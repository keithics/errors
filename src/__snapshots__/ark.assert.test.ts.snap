// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ark Error Handler Tests 404 Error Tests - should NOT return error 404 if value is NOT null,false or empty 1`] = `true`;

exports[`Ark Error Handler Tests 404 Error Tests - should NOT return error 404 if value is NOT null,false or empty 2`] = `true`;

exports[`Ark Error Handler Tests 404 Error Tests - should return error 404 if value is null.false or empty 1`] = `"404 Not Found"`;

exports[`Ark Error Handler Tests 404 Error Tests - should return error 404 if value is null.false or empty 2`] = `"404 Not Found"`;

exports[`Ark Error Handler Tests 404 Error Tests - should return error 404 if value is null.false or empty 3`] = `"404 Not Found"`;

exports[`Ark Error Handler Tests 500 Error Tests - should NOT return error 500 if value is NOT null,false or empty 1`] = `true`;

exports[`Ark Error Handler Tests 500 Error Tests - should NOT return error 500 if value is NOT null,false or empty 2`] = `true`;

exports[`Ark Error Handler Tests 500 Error Tests - should NOT return error 500 if value is a STRING or NUMBER 1`] = `true`;

exports[`Ark Error Handler Tests 500 Error Tests - should NOT return error 500 if value is a STRING or NUMBER 2`] = `true`;

exports[`Ark Error Handler Tests 500 Error Tests - should NOT return error 500 if value is a STRING or NUMBER 3`] = `true`;

exports[`Ark Error Handler Tests 500 Error Tests - should NOT return error 500 if value is a STRING or NUMBER 4`] = `true`;

exports[`Ark Error Handler Tests 500 Error Tests - should return error 500 if value is null.false or empty 1`] = `"An Error Occurred"`;

exports[`Ark Error Handler Tests 500 Error Tests - should return error 500 if value is null.false or empty 2`] = `"An Error Occurred"`;

exports[`Ark Error Handler Tests 500 Error Tests - should return error 500 if value is null.false or empty 3`] = `"An Error Occurred"`;

exports[`Ark Error Handler Tests Validation Error Tests - should not return an error if object is not empty or true 1`] = `true`;

exports[`Ark Error Handler Tests Validation Error Tests - should not return an error if object is not empty or true 2`] = `true`;

exports[`Ark Error Handler Tests Validation Error Tests - should return error if object is null, false or an empty object 1`] = `"Object is required"`;

exports[`Ark Error Handler Tests Validation Error Tests - should return error if object is null, false or an empty object 2`] = `"Object is required"`;

exports[`Ark Error Handler Tests Validation Error Tests - should return error if object is null, false or an empty object 3`] = `"Object is required"`;
